src_features/provideDomainName/cmd_provide_domain_name.c

```diff
@@ -0,0 +1,567 @@
+        {.tag = SIGNER_KEY_ID, .func = &handle_sign_key_id, .found = 0},
+        {.tag = SIGNER_ALGO, .func = &handle_sign_algo, .found = 0},
+        {.tag = SIGNATURE, .func = &handle_signature, .found = 0},
+        {.tag = DOMAIN_NAME, .func = &handle_domain_name, .found = 0},
+        {.tag = COIN_TYPE, .func = &handle_coin_type, .found = 0},
+        {.tag = ADDRESS, .func = &handle_address, .found = 0}};
+    e_tlv_step step = TLV_TAG;
+    s_tlv_data data;
+    uint16_t offset = 0;
+
+    cx_sha256_init(&sig_ctx->hash_ctx);
+    // handle TLV payload
+    while (offset < payload->size) {
+        switch (step) {
+            case TLV_TAG:
+                data.tag = payload->ptr[offset];
+                offset += sizeof(data.tag);
+                step = TLV_LENGTH;
+                break;
+
+            case TLV_LENGTH:
+                data.length = payload->ptr[offset];
+                offset += sizeof(data.length);
+                step = TLV_VALUE;
+                break;
+
+            case TLV_VALUE:
+                if (offset >= payload->size) {
+                    apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+                    return false;
+                }
+                data.value = &payload->ptr[offset];
+                // check if a handler exists for this tag
+                for (int idx = 0; idx < (int) ARRAY_SIZE(handlers); ++idx) {
+                    if (handlers[idx].tag == data.tag) {
+                        handlers[idx].found += 1;
+                        t_tlv_handler fptr = PIC(handlers[idx].func);
+                        if (!(*fptr)(&data, domain_name_info, sig_ctx)) {
+                            PRINTF("Error while handling tag 0x%x\n", handlers[idx].tag);
+                            apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+                            return false;
+                        }
+                        break;
+                    }
+                }
+                if (data.tag != SIGNATURE) {  // the signature wasn't computed on itself
+                    feed_tlv_hash(&data, sig_ctx);
+                }
+                offset += data.length;
+                step = TLV_TAG;
+                break;
+
+            default:
+                apdu_response_code = APDU_RESPONSE_INVALID_DATA;
+                return false;
+        }
+    }
+    // prevent missing or duplicated tags
+    for (int idx = 0; idx < (int) ARRAY_SIZE(handlers); ++idx) {
+        if (handlers[idx].found != 1) {
+            PRINTF("Found %u occurence(s) of tag 0x%x in TLV!\n",
+                   handlers[idx].found,
+                   handlers[idx].tag);
+            return false;
+        }
+    }
+    return true;
+}
+
+/**
+ * Allocate and assign TLV payload
+ *
+ * @param[in] payload payload structure
+ * @param[in] size size of the payload
+ * @return whether it was successful
+ */
+static bool set_payload(s_tlv_payload *payload, uint16_t size) {
```

Comment:
`allocate_payload` ?

Reply:
I changed it to something more meaningful. :+1: